### PR TITLE
Add bounds to allocations in advice map and keccak256/sha512 precompiles

### PR DESCRIPTION
**Pull request by plafer**
Addresses audit issue 36:

> Several host-side handlers derive allocation sizing and loop length from attacker-controlled sizes without explicit caps. In prover-service deployments where the attacker controls the program and advice inputs, adv.insert_mem, adv.push_mapval*, and the keccak256 precompile can be used to force large host-side CPU and memory usage that is not directly metered by VM cycle limits.

### CODE REVIEW

File: crates/lib/core/src/handlers/keccak256.rs [hunk 1]
```diff
@@ -67,6 +67,15 @@ impl EventHandler for KeccakPrecompile {
         let ptr = process.get_stack_item(1).as_canonical_u64();
         let len_bytes = process.get_stack_item(2).as_canonical_u64();
 
+        // Enforce the maximum precompile input size to prevent host-side resource exhaustion.
+        let max = process.execution_options().max_hash_len_bytes();
+        if len_bytes as usize > max {
```
Comment (huitseeker) [inline crates/lib/core/src/handlers/keccak256.rs]: This looks like a width cast before the limit check. On 32-bit targets, `len_bytes as usize` can truncate, so a value above `max_hash_len_bytes` may pass this guard and then flow into `read_memory_packed_u32` with the truncated length. Could we compare in `u64` first (`len_bytes > max as u64`) and then do a checked `usize::try_from(len_bytes)` before reading memory?

File: crates/lib/core/src/handlers/sha512.rs [hunk 2]
```diff
@@ -43,6 +43,15 @@ impl EventHandler for Sha512Precompile {
         let ptr = process.get_stack_item(1).as_canonical_u64();
         let len_bytes = process.get_stack_item(2).as_canonical_u64();
 
+        // Enforce the maximum precompile input size to prevent host-side resource exhaustion.
+        let max = process.execution_options().max_hash_len_bytes();
+        if len_bytes as usize > max {
```
Comment (huitseeker) [inline crates/lib/core/src/handlers/sha512.rs]: Same pattern here as in `keccak256`: the guard compares `len_bytes as usize` against `max`. On 32-bit this cast can truncate before the check. Would you be open to checking in `u64` first, then using `usize::try_from(len_bytes)` so the conversion cannot silently wrap?

File: processor/src/execution_options.rs [hunk 3]
```diff
@@ -38,6 +45,14 @@ impl ExecutionOptions {
     /// Default fragment size for core trace generation.
     pub const DEFAULT_CORE_TRACE_FRAGMENT_SIZE: usize = 4096; // 2^12
 
+    /// Default maximum number of field elements in a single advice map value inserted via
+    /// `adv.insert_mem`. Set to 2^17 (~1 MB given 8-byte field elements).
+    pub const DEFAULT_MAX_ADV_MAP_VALUE_SIZE: usize = 1 << 17;
+
+    /// Default maximum number of input bytes for a single hash precompile invocation (e.g.
+    /// keccak256, sha512, etc.). Set to 2^20 (1 MB).
+    pub const DEFAULT_MAX_HASH_LEN_BYTES: usize = 1 << 20;
```
Comment (huitseeker) [inline processor/src/execution_options.rs:54]: I may be missing a test, but I don’t see boundary coverage for the new caps through execution paths (`max` allowed, `max + 1` rejected) for both `max_hash_len_bytes` and `max_adv_map_value_size`. Could we add one focused test per boundary to lock this behavior in?

File: crates/lib/core/src/handlers/keccak256.rs (hunk 1 above)
Comment (plafer) [reply crates/lib/core/src/handlers/keccak256.rs]: Fixed

File: crates/lib/core/src/handlers/sha512.rs (hunk 2 above)
Comment (plafer) [reply crates/lib/core/src/handlers/sha512.rs]: Fixed

File: processor/src/execution_options.rs (hunk 3 above)
Comment (plafer) [reply processor/src/execution_options.rs:54]: Done

File: processor/src/host/advice/mod.rs [hunk 4]
```diff
@@ -120,16 +124,33 @@ impl AdviceProvider {
         Ok([word0, word1])
     }
 
+    /// Checks that pushing `count` elements would not exceed the advice stack size limit.
+    fn check_stack_capacity(&self, count: usize) -> Result<(), AdviceError> {
+        let resulting_size = self.stack.len() + count;
```
Comment (huitseeker) [inline processor/src/host/advice/mod.rs]: Not that this is likely, but could we make this addition overflow-safe with `checked_add`? If `len + count` wraps, the limit check can be bypassed before we return `StackSizeExceeded`.

File: processor/src/host/advice/mod.rs [hunk 5]
```diff
@@ -166,13 +187,19 @@ impl AdviceProvider {
     ) -> Result<(), AdviceError> {
         let values = self.map.get(&key).ok_or(AdviceError::MapKeyNotFound { key })?;
 
+        // Calculate total elements to push including padding and optional length prefix
+        let num_pad_elements = if pad_to != 0 {
+            values.len().next_multiple_of(pad_to as usize) - values.len()
+        } else {
+            0
+        };
+        let total_push = values.len() + num_pad_elements + if include_len { 1 } else { 0 };
```
Comment (huitseeker) [inline processor/src/host/advice/mod.rs]: Could we make `total_push` use `checked_add` as well? The pieces are attacker-influenced sizes, so wrapping here could undercount and skip the stack cap check.

File: processor/src/host/advice/mod.rs (hunk 4 above)
Comment (plafer) [reply processor/src/host/advice/mod.rs]: Done

File: processor/src/host/advice/mod.rs (hunk 5 above)
Comment (plafer) [reply processor/src/host/advice/mod.rs]: Done
